File: apps/server/src/ccsds.ts
// CCSDS Space Packet Protocol (133.0-B-2) + minimal PUS-C secondary header.
// Symmetric with apps/firmware/src/ccsds.{h,c}. Big-endian on the wire.
// See docs/PROTOCOL.md.

export const CCSDS_PRIMARY_HEADER_SIZE = 6;
export const CCSDS_PUS_TM_SEC_HDR_SIZE = 7;
export const CCSDS_PUS_TC_SEC_HDR_SIZE = 4;
export const CCSDS_CRC_SIZE = 2;

export const CCSDS_PVN = 0;
export const CCSDS_TYPE_TM = 0;
export const CCSDS_TYPE_TC = 1;
export const CCSDS_SEQF_UNSEGMENTED = 3;

export const PUS_VERSION_C = 0x10;

// APIDs — must match the firmware definitions.
export const APID = {
  HK: 0x100,
  EVENT: 0x101,
  ACK: 0x102,
  TC: 0x200,
} as const;

export const PUS = {
  VERIFICATION: 1,
  VERIF_ACCEPT_OK: 1,
  VERIF_ACCEPT_FAIL: 2,
  HOUSEKEEPING: 3,
  HK_REPORT: 25,
  EVENT: 5,
  EVENT_INFO: 1,
  EVENT_HIGH: 4,
  FUNCTION: 8,
  FUNCTION_CALL: 1,
  TEST: 17,
  TEST_REQUEST: 1,
  TEST_REPORT: 2,
} as const;

export const APOGEE_FN = {
  SET_MODE: 0x01,
  REBOOT: 0x02,
} as const;

// CRC-16-CCITT-FALSE: poly 0x1021, init 0xFFFF, no reflect, no xor-out.
export function crc16Ccitt(buf: Buffer | Uint8Array): number {
  let crc = 0xffff;
  for (let i = 0; i < buf.length; i++) {
    crc ^= (buf[i] as number) << 8;
    for (let b = 0; b < 8; b++) {
      crc = (crc & 0x8000) !== 0 ? ((crc << 1) ^ 0x1021) & 0xffff : (crc << 1) & 0xffff;
    }
  }
  return crc & 0xffff;
}

// --- Encoding ----------------------------------------------------------

export function packPrimaryHeader(
  buf: Buffer,
  offset: number,
  type: 0 | 1,
  secHdr: boolean,
  apid: number,
  seqCount: number,
  dataLengthMinus1: number,
): number {
  let w0 = (CCSDS_PVN & 0x07) << 13;
  w0 |= (type & 0x01) << 12;
  w0 |= (secHdr ? 1 : 0) << 11;
  w0 |= apid & 0x07ff;

  let w1 = (CCSDS_SEQF_UNSEGMENTED & 0x03) << 14;
  w1 |= seqCount & 0x3fff;

  buf.writeUInt16BE(w0, offset);
  buf.writeUInt16BE(w1, offset + 2);
  buf.writeUInt16BE(dataLengthMinus1 & 0xffff, offset + 4);
  return CCSDS_PRIMARY_HEADER_SIZE;
}

export function packPusTmSecondary(
  buf: Buffer,
  offset: number,
  service: number,
  subtype: number,
  cucSeconds: number,
): number {
  buf.writeUInt8(PUS_VERSION_C, offset);
  buf.writeUInt8(service, offset + 1);
  buf.writeUInt8(subtype, offset + 2);
  buf.writeUInt32BE(cucSeconds >>> 0, offset + 3);
  return CCSDS_PUS_TM_SEC_HDR_SIZE;
}

export function packPusTcSecondary(
  buf: Buffer,
  offset: number,
  ackFlags: number,
  service: number,
  subtype: number,
): number {
  buf.writeUInt8(PUS_VERSION_C, offset);
  buf.writeUInt8(ackFlags, offset + 1);
  buf.writeUInt8(service, offset + 2);
  buf.writeUInt8(subtype, offset + 3);
  return CCSDS_PUS_TC_SEC_HDR_SIZE;
}

export function appendCrc(buf: Buffer, packetLen: number): number {
  const crc = crc16Ccitt(buf.subarray(0, packetLen));
  buf.writeUInt16BE(crc, packetLen);
  return packetLen + CCSDS_CRC_SIZE;
}

// --- Decoding ----------------------------------------------------------

export type CcsdsPrimaryHeader = {
  version: number;
  type: 0 | 1;
  secHdrFlag: boolean;
  apid: number;
  seqFlags: number;
  seqCount: number;
  dataLength: number; // real bytes after PH = (PDL field) + 1
};

export type CcsdsPusTmSecondary = {
  pusVersion: number;
  service: number;
  subtype: number;
  cucSeconds: number;
};

export type CcsdsPusTcSecondary = {
  pusVersion: number;
  ackFlags: number;
  service: number;
  subtype: number;
};

export function parsePrimaryHeader(buf: Buffer): CcsdsPrimaryHeader | null {
  if (buf.length < CCSDS_PRIMARY_HEADER_SIZE) return null;
  const w0 = buf.readUInt16BE(0);
  const w1 = buf.readUInt16BE(2);
  const pdl = buf.readUInt16BE(4);
  const version = (w0 >> 13) & 0x07;
  if (version !== CCSDS_PVN) return null;
  const type = ((w0 >> 12) & 0x01) as 0 | 1;
  return {
    version,
    type,
    secHdrFlag: ((w0 >> 11) & 0x01) !== 0,
    apid: w0 & 0x07ff,
    seqFlags: (w1 >> 14) & 0x03,
    seqCount: w1 & 0x3fff,
    dataLength: pdl + 1,
  };
}

export function parsePusTmSecondary(buf: Buffer): CcsdsPusTmSecondary | null {
  if (buf.length < CCSDS_PUS_TM_SEC_HDR_SIZE) return null;
  const pusVersion = buf.readUInt8(0);
  if (pusVersion !== PUS_VERSION_C) return null;
  return {
    pusVersion,
    service: buf.readUInt8(1),
    subtype: buf.readUInt8(2),
    cucSeconds: buf.readUInt32BE(3),
  };
}

export function parsePusTcSecondary(buf: Buffer): CcsdsPusTcSecondary | null {
  if (buf.length < CCSDS_PUS_TC_SEC_HDR_SIZE) return null;
  const pusVersion = buf.readUInt8(0);
  if (pusVersion !== PUS_VERSION_C) return null;
  return {
    pusVersion,
    ackFlags: buf.readUInt8(1),
    service: buf.readUInt8(2),
    subtype: buf.readUInt8(3),
  };
}

export function verifyCrc(buf: Buffer): boolean {
  if (buf.length < CCSDS_CRC_SIZE) return false;
  const body = buf.length - CCSDS_CRC_SIZE;
  const got = buf.readUInt16BE(body);
  const expected = crc16Ccitt(buf.subarray(0, body));
  return got === expected;
}

File: apps/firmware/src/ccsds.c
#include "ccsds.h"
#include "crc16.h"

#include <string.h>

/* All multi-byte fields are big-endian (network order) per CCSDS. */

static void put_u16_be(uint8_t *p, uint16_t v) {
    p[0] = (uint8_t)(v >> 8);
    p[1] = (uint8_t)(v & 0xFFu);
}

static void put_u32_be(uint8_t *p, uint32_t v) {
    p[0] = (uint8_t)((v >> 24) & 0xFFu);
    p[1] = (uint8_t)((v >> 16) & 0xFFu);
    p[2] = (uint8_t)((v >> 8) & 0xFFu);
    p[3] = (uint8_t)(v & 0xFFu);
}

static uint16_t get_u16_be(const uint8_t *p) {
    return (uint16_t)((uint16_t)p[0] << 8) | (uint16_t)p[1];
}

size_t ccsds_pack_primary_header(uint8_t *buf, size_t buf_len,
                                 uint8_t type, bool sec_hdr,
                                 uint16_t apid, uint16_t seq_count,
                                 uint16_t data_length_minus_1) {
    if (buf_len < CCSDS_PRIMARY_HEADER_SIZE) return 0;

    /* word 0 : VER(3) | T(1) | S(1) | APID(11) */
    uint16_t w0 = (uint16_t)((CCSDS_PVN & 0x07u) << 13);
    w0 |= (uint16_t)((type & 0x01u) << 12);
    w0 |= (uint16_t)((sec_hdr ? 1u : 0u) << 11);
    w0 |= (uint16_t)(apid & 0x07FFu);

    /* word 1 : SEQF(2) | SEQ(14) */
    uint16_t w1 = (uint16_t)((CCSDS_SEQF_UNSEGMENTED & 0x03u) << 14);
    w1 |= (uint16_t)(seq_count & 0x3FFFu);

    put_u16_be(&buf[0], w0);
    put_u16_be(&buf[2], w1);
    put_u16_be(&buf[4], data_length_minus_1);
    return CCSDS_PRIMARY_HEADER_SIZE;
}

size_t ccsds_pack_pus_tm_secondary(uint8_t *buf, size_t buf_len,
                                   uint8_t service, uint8_t subtype,
                                   uint32_t cuc_seconds) {
    if (buf_len < CCSDS_PUS_TM_SEC_HDR_SIZE) return 0;
    buf[0] = PUS_VERSION_C;
    buf[1] = service;
    buf[2] = subtype;
    put_u32_be(&buf[3], cuc_seconds);
    return CCSDS_PUS_TM_SEC_HDR_SIZE;
}

size_t ccsds_pack_pus_tc_secondary(uint8_t *buf, size_t buf_len,
                                   uint8_t ack_flags,
                                   uint8_t service, uint8_t subtype) {
    if (buf_len < CCSDS_PUS_TC_SEC_HDR_SIZE) return 0;
    buf[0] = PUS_VERSION_C;
    buf[1] = ack_flags;
    buf[2] = service;
    buf[3] = subtype;
    return CCSDS_PUS_TC_SEC_HDR_SIZE;
}

size_t ccsds_append_crc(uint8_t *buf, size_t buf_len, size_t packet_len) {
    if (buf_len < packet_len + CCSDS_CRC_SIZE) return 0;
    uint16_t crc = crc16_ccitt(buf, packet_len);
    put_u16_be(&buf[packet_len], crc);
    return packet_len + CCSDS_CRC_SIZE;
}

bool ccsds_parse_primary_header(const uint8_t *buf, size_t buf_len,
                                CcsdsPrimaryHeader *out) {
    if (buf_len < CCSDS_PRIMARY_HEADER_SIZE) return false;
    uint16_t w0 = get_u16_be(&buf[0]);
    uint16_t w1 = get_u16_be(&buf[2]);
    uint16_t pdl = get_u16_be(&buf[4]);

    out->version       = (uint8_t)((w0 >> 13) & 0x07u);
    out->type          = (uint8_t)((w0 >> 12) & 0x01u);
    out->sec_hdr_flag  = ((w0 >> 11) & 0x01u) != 0u;
    out->apid          = (uint16_t)(w0 & 0x07FFu);
    out->seq_flags     = (uint8_t)((w1 >> 14) & 0x03u);
    out->seq_count     = (uint16_t)(w1 & 0x3FFFu);
    out->data_length   = (uint16_t)(pdl + 1u);
    return out->version == CCSDS_PVN;
}

bool ccsds_parse_pus_tc_secondary(const uint8_t *buf, size_t buf_len,
                                  CcsdsPusTcSecondaryHeader *out) {
    if (buf_len < CCSDS_PUS_TC_SEC_HDR_SIZE) return false;
    out->pus_version = buf[0];
    out->ack_flags   = buf[1];
    out->service     = buf[2];
    out->subtype     = buf[3];
    return out->pus_version == PUS_VERSION_C;
}

bool ccsds_verify_crc(const uint8_t *buf, size_t packet_len) {
    if (packet_len < CCSDS_CRC_SIZE) return false;
    size_t body = packet_len - CCSDS_CRC_SIZE;
    uint16_t expected = crc16_ccitt(buf, body);
    uint16_t got = get_u16_be(&buf[body]);
    return expected == got;
}
